cmd/buckets-scan-mgr.go

```diff
@@ -0,0 +1,278 @@
+}
+
+func (mgr *bucketsScanMgr) start() {
+	m := &sync.Mutex{}
+	c := sync.NewCond(m)
+
+	// A routine that discovers new buckets and initialize scan stats for each new bucket
+	go func() {
+		t := time.NewTimer(bucketsListInterval)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				buckets, err := mgr.bucketsLister(mgr.ctx, BucketOptions{})
+				if err == nil {
+					mgr.mu.Lock()
+					mgr.knownBuckets = make(map[string]struct{}, len(buckets))
+					for _, b := range buckets {
+						mgr.knownBuckets[b.Name] = struct{}{}
+					}
+					for bucket := range mgr.knownBuckets {
+						for _, set := range mgr.internal {
+							st := set[bucket]
+							if l, err := globalLifecycleSys.Get(bucket); err == nil && l.HasActiveRules("") {
+								st.lifecycle = true
+							}
+							set[bucket] = st
+						}
+					}
+					mgr.mu.Unlock()
+
+					m.Lock()
+					c.Broadcast()
+					m.Unlock()
+				}
+				t.Reset(bucketsListInterval)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Wait until first buckets listing is successful
+	m.Lock()
+	c.Wait() // Unlocks m, waits, then locks m again
+	m.Unlock()
+
+	// Clean up internal data when a deleted bucket is found
+	go func() {
+		const cleanInterval = 30 * time.Second
+
+		t := time.NewTimer(cleanInterval)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				mgr.mu.Lock()
+				for _, set := range mgr.internal {
+					for bkt := range set {
+						if _, ok := mgr.knownBuckets[bkt]; !ok {
+							delete(set, bkt)
+						}
+					}
+				}
+				mgr.mu.Unlock()
+
+				t.Reset(cleanInterval)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// A routine that sends the next bucket to scan for each erasure set listener
+	go func() {
+		tick := 10 * time.Second
+
+		t := time.NewTimer(tick)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				mgr.mu.RLock()
+				for id, ch := range mgr.bucketsCh {
+					if len(ch) == 0 {
+						b := mgr.unsafeGetNextBucket(id)
+						if b != "" {
+							select {
+							case ch <- b:
+							default:
+							}
+						}
+					}
+				}
+				mgr.mu.RUnlock()
+
+				t.Reset(tick)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// Return a channel of buckets names to scan a given erasure set identifier
+func (mgr *bucketsScanMgr) getBucketCh(id setID) chan string {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+
+	mgr.internal[id] = make(map[string]bucketScanStat)
+	mgr.bucketsCh[id] = make(chan string, 1)
+
+	return mgr.bucketsCh[id]
+}
+
+func scanBefore(st1, st2 bucketScanStat) bool {
+	if st1.ongoing != st2.ongoing {
+		return st1.ongoing == false
+	}
+	if st1.lastFinished.Before(st2.lastFinished) {
+		return true
+	}
+	if st1.lifecycle != st2.lifecycle {
+		return st1.lifecycle == true
+	}
+	return false
+}
+
+// Return the next bucket name to scan of a given erasure set identifier
+// If all buckets are in a scanning state, return empty result
+func (mgr *bucketsScanMgr) unsafeGetNextBucket(id setID) string {
```

Comment:
I prefer to use "unsafe" for memory safety. IMO this more clearly signifies that the caller has locked m.
```suggestion
func (mgr *bucketsScanMgr) getNextBucketLocked(id setID) string {
```